File: wms/public/js/sales_invoice.js
frappe.ui.form.on("Sales Invoice", {
    refresh: function(frm) {
        if (frm.doc.docstatus === 1) {
            frm.add_custom_button(
                __('Pick List'),
                () => frm.trigger('create_pick_list'),
                __('Create')
            );
        }
    },
    create_pick_list: function(frm) {
        frappe.model.open_mapped_doc({
            frm,
            method: "wms.custom.sales_invoice.create_pick_list"
        });
    }
});

File: wms/custom/sales_invoice.py
import frappe
from frappe.utils import flt
from frappe.model.mapper import get_mapped_doc


@frappe.whitelist()
def create_pick_list(source_name, target_doc=None):
    def update_item_quantity(source, target, source_parent):
        target.qty = flt(source.qty) - flt(source.delivered_qty)
        target.stock_qty = (flt(source.qty) - flt(source.delivered_qty)) * flt(source.conversion_factor)

    doc = get_mapped_doc('Sales Invoice', source_name, {
        'Sales Invoice': {
            'doctype': 'Pick List',
            'validation': {
                'docstatus': ['=', 1]
            }
        },
        'Sales Invoice Item': {
            'doctype': 'Pick List Item',
            'field_map': {
                'parent': 'wms_sales_invoice',
                'name': 'wms_sales_invoice_item'
            },
            'postprocess': update_item_quantity
        }
    }, target_doc)

    doc.purpose = 'Delivery against Sales Invoice'

    return doc
